code/replication/object/Procedure.py

```diff
@@ -0,0 +1,76 @@
+from Similarity import *
+from BSBManager import *
+from Article import Article
+import time
+import glob
+import sys
+from multiprocessing import cpu_count
+import csv
+from ETUtils import *
+
+def procedure(startlocation = 'data', endlocation='export_dataframe.csv', simtest=CosineSimilarity(), worker_count=-1):
+	'''
+	Performs the procedure for the specified amount of articles. Uses 
+	all nml files from startlocation, and exports a csv file at endlocation.
+	'''
+
+	if worker_count < 0:
+		worker_count += cpu_count() + 1
+
+	worker_count = worker_count * 3
+	
+	location = sorted(glob.glob(startlocation + '/*.nml'))
+	companies = dict()
+	suppliers, supplier_processes = worker_init(worker_count, "supplier", simtest)
+	processors, processor_processes = worker_init(worker_count, "processor", simtest)
+	for f in location:
+		print("File processing...",f)
+		xtg = textGetter(f)
+		
+		for supplier in suppliers: #Send out the first batch of articles
+			try:
+				et = next(xtg)
+			except StopIteration:
+				continue
+			supplier.send(et)
+
+		checks, load = 0, 0
+		while checks < len(suppliers): # Makes sure to get back all articles before moving on
+			for supplier in suppliers:
+				if checks >= len(suppliers):
+					break
+
+				story = supplier.recv() # Gets back an article in apropriate oreder by waiting
+
+				try:
+					et = next(xtg)
+					supplier.send(et)
+				except StopIteration:
+					checks += 1 
+				
+				if not story.bad and not (story.tickers == []):
+					for ticker in story.tickers:
+						if '.' in ticker:
+							continue
+						if ticker not in companies:
```

Comment:
Please add inline documentation for round-robin ticker-process invariant. Very neat, but many people won't get it whe reading the code--i.e. it's good to know this is happening before you read the outer loop...